<a href="https://colab.research.google.com/github/smonodeep/esnkhyaki-mospi/blob/main/esankhyiki_tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# esankhyiki - Examples

Python client for India's National Statistical Office (MoSPI) data portal.
Access indicators across 22 datasets: employment, prices, GDP, health, education, environment, and more.

```bash
pip install mospi-esankhyiki[pandas]
```

In [32]:
pip install mospi-esankhyiki

In [33]:
import warnings
warnings.filterwarnings('ignore')
import esankhyiki
from esankhyiki.exceptions import InvalidDatasetError, InvalidFilterError, APIError, NoDataError

---
## Available Datasets

In [34]:
esankhyiki.list_datasets(format="df")

,dataset
0,PLFS
1,CPI
2,IIP
3,ASI
4,NAS
5,WPI
6,ENERGY
7,AISHE
8,ASUSE
9,GENDER


---
## PLFS - Periodic Labour Force Survey

Covers labour force participation, unemployment, worker population ratio, wages.
Use `get_indicators` to see what indicator codes are available, `get_metadata` to see valid filter values for a given indicator.

In [35]:
# Indicators grouped by frequency_code
indicators = esankhyiki.get_indicators("PLFS")
for group, items in indicators["indicators_by_frequency"].items():
    print(f"\n{group}:")
    for item in items:
        print(f"  {item['indicator_code']}: {item.get('indicator_name', item.get('description', ''))}")


frequency_code_1_Annual:
  1: LFPR (Labour Force Participation Rate, in per cent)
  2: WPR (Worker Population Ratio, in per cent)
  3: UR (Unemployment Rate, in per cent)
  4: Percentage distribution of workers
  5: Percentage of regular wage/ salaried employees with employment condition as
  6: Average wage/salary earnings (Rs. 0.00) during the preceding calendar month from regular wage/salaried employment
  7: Average wage earnings (Rs. 0.00) per day from casual labour work other than public works
  8: Average gross earnings (Rs. 0.00) during last 30 days from self-employment

frequency_code_2_Quarterly:
  1: LFPR (Labour Force Participation Rate, in per cent)
  2: WPR (Worker Population Ratio, in per cent)
  3: UR (Unemployment Rate, in per cent)
  4: Percentage distribution of workers

frequency_code_3_Monthly:
  1: LFPR (Labour Force Participation Rate, in per cent)
  2: WPR (Worker Population Ratio, in per cent)
  3: UR (Unemployment Rate, in per cent)


In [36]:
# Valid filter values for Unemployment Rate (indicator_code=3), Annual (frequency_code=1)
metadata = esankhyiki.get_metadata("PLFS", indicator_code=3, frequency_code=1, year_type_code=1)
for key, values in metadata["filter_values"]["data"].items():
    print(f"\n{key}:")
    for v in values[:4]:
        print(f"  {v}")


year_type:
  {'description': 'Calendar Year', 'year_type_code': 2}
  {'description': 'Agriculture Year', 'year_type_code': 1}

year:
  {'year': '2025'}
  {'year': '2024'}
  {'year': '2023-24'}
  {'year': '2023'}

age:
  {'age_code': 1, 'description': '15 years and above'}
  {'age_code': 2, 'description': '15-29 years'}
  {'age_code': 3, 'description': '15-59 years'}
  {'age_code': 4, 'description': 'all'}

education:
  {'description': 'all', 'education_code': 0}
  {'description': '1.not literate', 'education_code': 1}
  {'description': '2.literate & upto primary', 'education_code': 2}
  {'description': '3.middle', 'education_code': 3}

gender:
  {'description': 'male', 'gender_code': 1}
  {'description': 'female', 'gender_code': 2}
  {'description': 'person', 'gender_code': 3}

religion:
  {'description': 'all', 'religion_code': 1}
  {'description': 'Hinduism', 'religion_code': 2}
  {'description': 'Islam', 'religion_code': 3}
  {'description': 'Christianity', 'religion_code': 4}

sec

In [37]:
# Unemployment Rate - All India, all years, both sectors
esankhyiki.get_data("PLFS", {
    "indicator_code": 3,
    "frequency_code": 1,
    "year_type_code":1,
    "state_code": 99,     # 99 = All India
    "gender_code": 3,     # 3 = Person
    "age_code": 1,        # 1 = 15 years and above
    "sector_code": 3,     # 3 = rural + urban
}, format="df")

,year_type,year,frequency,indicator,state,gender,sector,AgeGroup,weekly_status,religion,socialGroup,General_Education,quarter,month,value,unit
0,Agriculture Year,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,CWS,all,all,all,all,None,8.7,%
1,Agriculture Year,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,all,all,None,6.0,%
2,Agriculture Year,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,1.not literate,all,None,1.2,%
3,Agriculture Year,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,2.literate & upto primary,all,None,2.7,%
4,Agriculture Year,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,3.middle,all,None,5.5,%
5,Agriculture Year,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,4.secondary,all,None,5.7,%
6,Agriculture Year,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,5.higher secondary,all,None,10.3,%
7,Agriculture Year,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,6.diploma/ certificate course,all,None,19.8,%
8,Agriculture Year,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,7.graduate,all,None,17.2,%
9,Agriculture Year,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,8.post graduate & above,all,None,14.6,%


---
## CPI - Consumer Price Index

Multiple base years (2010, 2012, 2024) and two levels: Group and Item.

In [38]:
# Available base years, levels, and series
esankhyiki.get_indicators("CPI", format="df")

,base_year,level,series
0,2010,Group,Current
1,2010,Group,Back
2,2010,Item,Current
3,2010,Item,Back
4,2012,Group,Current
5,2012,Group,Back
6,2012,Item,Current
7,2012,Item,Back
8,2024,Group,Current
9,2024,Group,Back


In [39]:
# Filter values for base_year=2012, Group level
metadata = esankhyiki.get_metadata("CPI", base_year="2012", level="Group", series="Current", format="df")
metadata

,series,year,month_code,month_name,state_code,state_name,sector_code,sector_name,group_code,group_name,subgroup_code,subgroup_name
0,Current,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Back,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,2025.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,NaN,2024.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,NaN,2023.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
100,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6,NaN,6.1.04,Recreation and Amusement
101,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6,NaN,6.1.05,Education
102,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6,NaN,6.1.06,Personal Care and Effects
103,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6,NaN,6.99,Miscellaneous-Overall


In [40]:
# CPI Group index, base year 2012, year 2024
esankhyiki.get_data("CPI", {
    "base_year": "2012",
    "series": "Current",
    "year": "2024",
}, format="df")

,baseyear,year,month,state,sector,group,subgroup,index,inflation,status
0,2012,2024,December,All India,Rural,Consumer Food Price,Consumer Food Price-Overall,204.7,8.65,F
1,2012,2024,December,All India,Rural,Miscellaneous,Transport and Communication,176.7,2.85,F
2,2012,2024,December,All India,Rural,Miscellaneous,Education,192.2,3.72,F
3,2012,2024,December,All India,Rural,Miscellaneous,Miscellaneous-Overall,190.8,4.26,F
4,2012,2024,December,All India,Rural,Miscellaneous,Household Goods and Services,187.0,2.47,F
5,2012,2024,December,All India,Rural,Miscellaneous,Health,200.2,4.00,F
6,2012,2024,December,All India,Rural,Miscellaneous,Personal Care and Effects,206.3,9.68,F
7,2012,2024,December,All India,Rural,Miscellaneous,Recreation and Amusement,181.5,2.54,F
8,2012,2024,December,All India,Rural,Fuel and Light,Fuel and Light-Overall,182.3,-0.44,F
9,2012,2024,December,All India,Rural,Housing,Housing-Overall,None,None,F


---
## NAS - National Accounts Statistics (GDP)

Annual and quarterly GDP, GVA, consumption, capital formation by industry and institutional sector.

In [41]:
esankhyiki.get_indicators("NAS", format="df")

,indicator_code,description
0,1,Gross Value Added
1,2,Net Taxes on Products
2,3,Taxes on Products
3,4,Subsidies on Products
4,5,Gross Domestic Product
5,6,Consumption of Fixed Capital
6,7,Net Domestic Product
7,8,Gross Capital Formation by Industry of Use
8,9,Gross Fixed Capital Formation
9,10,Private Final Consumption Expenditure


In [42]:
# Filter values for GVA by industry (indicator_code=1), base year 2022-23
metadata = esankhyiki.get_metadata("NAS", indicator_code=1, base_year="2022-23", series="Current", frequency_code=1)
for key, values in metadata.items():
    if isinstance(values, list):
        print(f"{key}: {values[:3]}")

year: [{'year': '2025-26'}, {'year': '2024-25'}, {'year': '2023-24'}]
revision: [{'revision_code': 2, 'revision_name': 'Second Advance Estimates'}, {'revision_code': 4, 'revision_name': 'First Revised Estimates'}, {'revision_code': 8, 'revision_name': 'Final Estimates'}]
industry: [{'industry_code': 1, 'industry_name': 'Agriculture, Livestock, Forestry and Fishing'}, {'industry_code': 2, 'industry_name': 'Mining and Quarrying'}, {'industry_code': 3, 'industry_name': 'Manufacturing'}]
subindustry: [{'subindustry_code': 1, 'subindustry_name': 'Crops', 'industry_code': 1}, {'subindustry_code': 2, 'subindustry_name': 'Livestock', 'industry_code': 1}, {'subindustry_code': 3, 'subindustry_name': 'Forestry and Logging', 'industry_code': 1}]
institutional_sector: [{'institutional_code': 1, 'institutional_name': 'Public Non-Financial Corporations'}, {'institutional_code': 2, 'institutional_name': 'Private Non-Financial Corporations'}, {'institutional_code': 3, 'institutional_name': 'Public Fina

In [43]:
# GVA by industry, current prices, latest estimates
esankhyiki.get_data("NAS", {
    "indicator_code": 1,
    "base_year": "2022-23",
    "series": "Current",
    "frequency_code": 1,
    "year": "2024-25",
}, format="df")

,base_year,series,year,indicator,frequency,revision,industry,subindustry,institutional_sector,quarter,current_price,constant_price,unit
0,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Public Non-Financial Corporations,None,1551904.3111790416,None,₹ Crore
1,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Private Non-Financial Corporations,None,9555601.402799515,None,₹ Crore
2,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Public Financial Corporations,None,780342.4093269374,None,₹ Crore
3,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Private Financial Corporations,None,1059339.3228545364,None,₹ Crore
4,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,General Government,None,2656799.960246522,None,₹ Crore
5,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Households Including NPISH,None,13250480.021178257,None,₹ Crore
6,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Total Gross Value Added,None,28854467.42758481,None,₹ Crore
7,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,"Agriculture, Livestock, Forestry and Fishing",Crops,None,None,3487514.5392785612,3113280.9316670736,₹ Crore
8,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,"Agriculture, Livestock, Forestry and Fishing",Livestock,None,None,1387062.690777359,1278634.3400525874,₹ Crore
9,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,"Agriculture, Livestock, Forestry and Fishing",Forestry and Logging,None,None,408682.0267620707,414487.83170621865,₹ Crore


---
## IIP - Index of Industrial Production

Monthly and annual industrial output across manufacturing, mining, and electricity sectors.

In [44]:
# Available categories and years for IIP annual
metadata = esankhyiki.get_metadata("IIP", base_year="2011-12", frequency="Annually")
metadata

{'financial_year': [{'financial_year': '2025-26'},
  {'financial_year': '2024-25'},
  {'financial_year': '2023-24'},
  {'financial_year': '2022-23'},
  {'financial_year': '2021-22'},
  {'financial_year': '2020-21'},
  {'financial_year': '2019-20'},
  {'financial_year': '2018-19'},
  {'financial_year': '2017-18'},
  {'financial_year': '2016-17'},
  {'financial_year': '2015-16'},
  {'financial_year': '2014-15'},
  {'financial_year': '2013-14'},
  {'financial_year': '2012-13'}],
 'type': [{'type': 'Use-based category'},
  {'type': 'Sectoral'},
  {'type': 'General'}],
 'category': [{'category_code': 1,
   'category_name': 'Mining',
   'type': 'Sectoral'},
  {'category_code': 2, 'category_name': 'Manufacturing', 'type': 'Sectoral'},
  {'category_code': 3, 'category_name': 'Electricity', 'type': 'Sectoral'},
  {'category_code': 4, 'category_name': 'General', 'type': 'General'},
  {'category_code': 5,
   'category_name': 'Primary Goods',
   'type': 'Use-based category'},
  {'category_code': 6

In [45]:
# IIP annual, 2023-24
esankhyiki.get_data("IIP", {
    "base_year": "2011-12",
    "financial_year": "2023-24",
}, format="df")

,base_year,year,type,category,sub_category,index,growth_rate
0,2011-12,2023-24,General,General,,146.7,5.9
1,2011-12,2023-24,Sectoral,Mining,,128.9,7.5
2,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Food Products,134.5,1.6
3,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Beverages,110.9,5.1
4,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Tobacco Products,81.1,-8.3
5,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Textiles,107.6,0.1
6,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Wearing Apparel,109.9,-14.1
7,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Leather and Related Products,95.0,-1.0
8,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Wood and Products of Wood and C...,98.3,-5.9
9,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Paper and Paper Products,79.4,-3.6


---
## GENDER - Gender Statistics

147 indicators covering demographics, health, education, labour, financial inclusion, and crimes against women.

In [46]:
indicators = esankhyiki.get_indicators("GENDER")
print(f"Total indicators: {len(indicators)}\n")
for i in indicators[:10]:
    print(f"  {i['indicator_code']}: {i['description']}")

Total indicators: 147

  1: Trend in Population
  2: Sex Ratio
  3: Average Annual Growth Rate of Population
  4: Percentage Distribution of Projected Population
  5: Sex ratio in context of developmental planning
  6: State-wise Sex Ratio 
  7: Sex ratio at birth for bigger States based on Sample Registration System
  8: General Fertility Rate
  9: Age-specific Fertility rate
  10: Total Fertility Rate 


In [47]:
# Sex ratio (indicator_code=2)
esankhyiki.get_data("GENDER", {"indicator_code": 2}, format="df")

,year,state/UT,indicator,value,unit
0,1951,All India,Sex Ratio,946,%
1,1951,Andaman & Nicobar Islands,Sex Ratio,625,%
2,1951,Andhra Pradesh,Sex Ratio,986,%
3,1951,Arunachal Pradesh,Sex Ratio,-,%
4,1951,Assam,Sex Ratio,868,%
5,1951,Bihar,Sex Ratio,1000,%
6,1951,Chandigarh,Sex Ratio,781,%
7,1951,Chhattisgarh,Sex Ratio,1024,%
8,1951,Dadra & Nagar Haveli,Sex Ratio,946,%
9,1951,Daman & Diu,Sex Ratio,1125,%


---
## ENVSTATS - Environment Statistics

124 indicators covering climate, water, forests, biodiversity, pollution, and natural disasters.

In [48]:
indicators = esankhyiki.get_indicators("ENVSTATS")
print(f"Total indicators: {len(indicators)}\n")
for i in indicators[:10]:
    print(f"  {i['indicator_code']}: {i.get('description', i.get('label', ''))}")

Total indicators: 124

  1: Annual and Seasonal Mean Temperature
  2: Annual and Seasonal Minimum and Maximum Temperature
  3: Annual and Monthly Rainfall
  4: Wetland distribution in India
  5: Number and Area of Wetlands
  6: Water Sheds in India
  7: Catchment Area of major River Basins
  8: Length of Coastline and Coastal Population of India
  9: Occurrence of Major Soils in India
  10: Category-wise distribution of land degradation in different States of India


In [49]:
# Mean temperature trend (indicator_code=1)
esankhyiki.get_data("ENVSTATS", {"indicator_code": 1}, format="df")

,year,indicator,period,value,unit
0,2024,Annual and Seasonal Mean Temperature,Annual,25.7431,°C
1,2024,Annual and Seasonal Mean Temperature,Jan-Feb,19.7385,°C
2,2024,Annual and Seasonal Mean Temperature,Mar-May,28.3522,°C
3,2024,Annual and Seasonal Mean Temperature,June-Sept,28.7166,°C
4,2024,Annual and Seasonal Mean Temperature,Oct-Dec,23.1724,°C
...,...,...,...,...,...
115,2001,Annual and Seasonal Mean Temperature,Annual,25.0501,°C
116,2001,Annual and Seasonal Mean Temperature,Jan-Feb,19.4017,°C
117,2001,Annual and Seasonal Mean Temperature,Mar-May,27.7716,°C
118,2001,Annual and Seasonal Mean Temperature,June-Sept,27.7377,°C


---
## RBI - Foreign Trade and Exchange Rates

39 indicators on balance of payments, forex reserves, exchange rates (155 currencies), NRI deposits.

In [50]:
indicators = esankhyiki.get_indicators("RBI")
for i in indicators:
    print(f"  {i['indicator_code']}: {i['label']}")

  22: Balance of Payments ‑ Indicators
  24: Broad Commodity Composition of India's Merchandise Trade ‑ Oil and Non oil Exports and Imports ‑ Rupees
  42: Broad Commodity Composition of India’s Merchandise Trade - Oil and Non-Oil Exports and Imports - US Dollar
  1: Direction of Foreign Trade ‑ Rupees
  11: Direction of Foreign Trade ‑ US Dollars
  32: Exchange Rate of the Indian Rupee Vis‑a‑Vis the SDR, US Dollar, Pound Sterling, D.M./Euro and Japanese Yen (Calendar Year ‑ Annual Average)
  31: Exchange Rate of the Indian Rupee Vis‑A‑Vis the SDR, US Dollar, Pound Sterling, D.M./Euro and Japanese Yen (Financial Year ‑ Annual Average and End‑Year Rates)
  35: Exchange Rate of the Indian Rupee vis-à-vis the SDR, US Dollar, Pound Sterling, Euro and Japanese Yen (High and Low During the Month)
  33: Exchange Rate of the Indian Rupee vis-à-vis the SDR, US Dollar, Pound Sterling, Euro and Japanese Yen (Monthly Average and end-Month Rates)
  43: EXPORTS OF MAJOR COMMODITIES - RUPEES
  44: EXP

In [51]:
# Get valid sub_indicator_codes for Balance of Payments (indicator_code=22)
metadata = esankhyiki.get_metadata("RBI", sub_indicator_code=22)
metadata

[{'sub_indicator_code': 22,
  'label': 'Balance of Payments ‑ Indicators',
  'dimensions': {'year': [{'id': '2024-25', 'label': '2024-25'},
    {'id': '2023-24', 'label': '2023-24'},
    {'id': '2022-23', 'label': '2022-23'},
    {'id': '2021-22', 'label': '2021-22'},
    {'id': '2020-21', 'label': '2020-21'},
    {'id': '2019-20', 'label': '2019-20'},
    {'id': '2018-19', 'label': '2018-19'},
    {'id': '2017-18', 'label': '2017-18'},
    {'id': '2016-17', 'label': '2016-17'},
    {'id': '2015-16', 'label': '2015-16'},
    {'id': '2014-15', 'label': '2014-15'},
    {'id': '2013-14', 'label': '2013-14'},
    {'id': '2012-13', 'label': '2012-13'},
    {'id': '2011-12', 'label': '2011-12'},
    {'id': '2010-11', 'label': '2010-11'},
    {'id': '2009-10', 'label': '2009-10'},
    {'id': '2008-09', 'label': '2008-09'},
    {'id': '2007-08', 'label': '2007-08'},
    {'id': '2006-07', 'label': '2006-07'},
    {'id': '2005-06', 'label': '2005-06'},
    {'id': '2004-05', 'label': '2004-05'},


In [52]:
# Balance of Payments
esankhyiki.get_data("RBI", {"indicator_code": 22}, format="df")

,indicator,year,balance_of_payments,trade_category,value,unit
0,Balance of Payments ‑ Indicators,2024-25,Trade,Exports/ GDP,11.30,%
1,Balance of Payments ‑ Indicators,2024-25,Trade,Imports/ GDP,18.65,%
2,Balance of Payments ‑ Indicators,2024-25,Invisibles,Receipts/GDP,14.75,%
3,Balance of Payments ‑ Indicators,2024-25,Invisibles,Payments/GDP,8.00,%
4,Balance of Payments ‑ Indicators,2024-25,Invisibles,Net/ GDP,6.75,%
5,Balance of Payments ‑ Indicators,2024-25,Current account,Current Receipts/ GDP,26.04,%
6,Balance of Payments ‑ Indicators,2024-25,Current account,Current Receipts/ Current Payments,97.76,%
7,Balance of Payments ‑ Indicators,2024-25,Capital account,Foreign Investment/ Exports,1.02,%
8,Balance of Payments ‑ Indicators,2024-25,Capital account,Foreign Investment/ GDP,0.12,%
9,Balance of Payments ‑ Indicators,2024-25,Import cover of Reserves (in months),None,11.00,%


---
## EC - Economic Census

District-level establishment and worker counts across three census editions (EC4/EC5/EC6).

In [53]:
esankhyiki.get_indicators("EC", format="df")

,indicator_code,indicator_name,description
0,1,Sixth Economic Census (EC6) - 2013-14,District-wise establishment and worker counts....
1,2,Fifth Economic Census (EC5) - 2005,District-wise establishment and worker counts....
2,3,Fourth Economic Census (EC4) - 1998,District-wise establishment and worker counts....


In [54]:
# EC6 (2013-14) - available states and filter options
metadata = esankhyiki.get_metadata("EC", indicator_code=1)
metadata

{'state': [{'id': '01', 'name': 'Jammu And Kashmir'},
  {'id': '02', 'name': 'Himachal Pradesh'},
  {'id': '03', 'name': 'Punjab'},
  {'id': '04', 'name': 'Chandigarh'},
  {'id': '05', 'name': 'Uttarakhand'},
  {'id': '06', 'name': 'Haryana'},
  {'id': '07', 'name': 'Nct Of Delhi'},
  {'id': '08', 'name': 'Rajasthan'},
  {'id': '09', 'name': 'Uttar Pradesh'},
  {'id': '10', 'name': 'Bihar'},
  {'id': '11', 'name': 'Sikkim'},
  {'id': '12', 'name': 'Arunachal Pradesh'},
  {'id': '13', 'name': 'Nagaland'},
  {'id': '14', 'name': 'Manipur'},
  {'id': '15', 'name': 'Mizoram'},
  {'id': '16', 'name': 'Tripura'},
  {'id': '17', 'name': 'Meghalaya'},
  {'id': '18', 'name': 'Assam'},
  {'id': '19', 'name': 'West Bengal'},
  {'id': '20', 'name': 'Jharkhand'},
  {'id': '21', 'name': 'Orissa'},
  {'id': '22', 'name': 'Chhattisgarh'},
  {'id': '23', 'name': 'Madhya Pradesh'},
  {'id': '24', 'name': 'Gujarat'},
  {'id': '25', 'name': 'Daman And Diu'},
  {'id': '26', 'name': 'Dadra And Nagar Haveli'

In [55]:
# EC6 - top districts in Maharashtra by establishments
esankhyiki.get_data("EC", {
    "indicator_code": 1,
    "state": "27",       # Maharashtra
    "mode": "ranking",
    "top5opt": "2",
}, format="df")

,rank,district,establishments
0,1,Kolhapur,570867
1,2,Pune,514469
2,3,Mumbai Suburban,465416
3,4,Thane,438706
4,5,Satara,383647
5,6,Solapur,373286
6,7,Sangli,278060
7,8,Mumbai,261259
8,9,Nagpur,258711
9,10,Ahmadnagar,246417


---
## MNRE - Renewable Energy

State-wise monthly installed capacity (in MW) for solar, wind, hydro, bio, and total renewable power from the Ministry of New and Renewable Energy.

In [56]:
# Energy types: 1=Solar, 2=Wind, 3=Hydro, 4=Bio, 5=Total
# Solar/Hydro/Bio support category_code breakdown; Wind and Total do not
esankhyiki.get_indicators("MNRE", format="df")

,type_of_renewable_energy_code,label
0,1,Solar Power
1,2,Wind Power
2,3,Hydro Power
3,4,Bio Power
4,5,Total Power


In [57]:
# Filter values for Solar Power (type 1) - returns valid category_codes
metadata = esankhyiki.get_metadata("MNRE", indicator_code=1)
metadata

[{'year': [{'year': 2026},
   {'year': 2025},
   {'year': 2024},
   {'year': 2023},
   {'year': 2022},
   {'year': 2021},
   {'year': 2020}],
  'month': [{'month_code': 1, 'label': 'January'},
   {'month_code': 2, 'label': 'February'},
   {'month_code': 3, 'label': 'March'},
   {'month_code': 4, 'label': 'April'},
   {'month_code': 5, 'label': 'May'},
   {'month_code': 6, 'label': 'June'},
   {'month_code': 7, 'label': 'July'},
   {'month_code': 8, 'label': 'August'},
   {'month_code': 9, 'label': 'September'},
   {'month_code': 10, 'label': 'October'},
   {'month_code': 11, 'label': 'November'},
   {'month_code': 12, 'label': 'December'}],
  'state': [{'state_code': 36, 'label': 'All India'},
   {'state_code': 39, 'label': 'Andaman and Nicobar Islands'},
   {'state_code': 1, 'label': 'Andhra Pradesh'},
   {'state_code': 2, 'label': 'Arunachal Pradesh'},
   {'state_code': 3, 'label': 'Assam'},
   {'state_code': 4, 'label': 'Bihar'},
   {'state_code': 30, 'label': 'Chandigarh'},
   {'st

In [58]:
esankhyiki.get_data("MNRE", {
    "type_of_renewable_energy_code": 1,  # Solar Power
    "state_code": 36,                    # All India
    "year": "2023",
}, format="df")

,type_of_renewable_energy,year,month,state,category,value
0,Solar Power,2023,December,All India,Ground Mounted Solar Power,56920.20
1,Solar Power,2023,December,All India,Rooftop Solar Power,11078.94
2,Solar Power,2023,December,All India,Hybrid Solar Component,2571.96
3,Solar Power,2023,December,All India,Off-Grid Solar Power/KUSUM,2748.39
4,Solar Power,2023,December,All India,Solar Power Total,73319.50
5,Solar Power,2023,November,All India,Ground Mounted Solar Power,55958.18
6,Solar Power,2023,November,All India,Rooftop Solar Power,11078.94
7,Solar Power,2023,November,All India,Hybrid Solar Component,2570.96
8,Solar Power,2023,November,All India,Off-Grid Solar Power/KUSUM,2703.53
9,Solar Power,2023,November,All India,Solar Power Total,72311.62


---
## Output Formats

All functions support `format="dict"` (default), `format="df"` (pandas DataFrame), or `format="csv"`.

In [59]:
csv_str = esankhyiki.get_data("PLFS", {
    "indicator_code": 3,
    "frequency_code": 1,
    "year_type_code":1,
    "year": "2023-24",
    "state_code": 99,
    "gender_code": 3,
    "age_code": 1,
    "sector_code": 3,
}, format="csv")

print(csv_str[:600])

year_type,year,frequency,indicator,state,gender,sector,AgeGroup,weekly_status,religion,socialGroup,General_Education,quarter,month,value,unit
Agriculture Year,2023-24,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,CWS,all,all,all,all,,4.9,%
Agriculture Year,2023-24,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,all,all,,3.2,%
Agriculture Year,2023-24,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,1.not literate,all,


---
## Error Handling

In [60]:
# Typo in dataset name - fuzzy suggestion
try:
    esankhyiki.get_indicators("plfs_data")
except InvalidDatasetError as e:
    print(e)

Unknown dataset: 'plfs_data'. Valid datasets: PLFS, CPI, IIP, ASI, NAS, WPI, ENERGY, AISHE, ASUSE, GENDER, NFHS, ENVSTATS, RBI, NSS77, NSS78, CPIALRL, HCES, TUS, EC, NSS79, UDISE, MNRE. Did you mean: PLFS, ENVSTATS, NFHS?


In [61]:
# Missing required parameter
try:
    esankhyiki.get_metadata("PLFS", frequency_code=1)
except InvalidFilterError as e:
    print(e)

Missing required params for get_metadata(dataset='PLFS'): ['indicator_code', 'year_type_code']


In [62]:
# Invalid filter parameter name
try:
    esankhyiki.get_data("PLFS", {
        "indicator_code": 1,
        "frequency_code": 1,
        "typo_param": "hello",
    })
except InvalidFilterError as e:
    print(e)

Invalid params: ['typo_param']. Valid params: ['indicator_code', 'frequency_code', 'year_type_code', 'year', 'month_code', 'state_code', 'gender_code', 'age_code', 'sector_code', 'quarter_code', 'weekly_status_code', 'religion_code', 'social_category_code', 'education_code', 'broad_industry_work_code', 'broad_status_employment_code', 'employee_contract_code', 'enterprise_size_code', 'enterprise_type_code', 'industry_section_code', 'nco_division_code', 'nic_group_code', 'limit', 'page', 'Format'].
